OasisDEX/oasis-borrow
# Cache & batch Infura calls on Server

# [Off-chain cache for Infura calls](https://app.shortcut.com/oazo-apps/story/5172/off-chain-cache-for-infura-calls)

Reduces infura calls for concurrent sessions. Current TTL set at 15 seconds but can be increased to further reduce infura call volume at the cost of slightly stale data in app.
  
## Changes 👷‍♀️

- Add new custom JsonRpcProvider called `JsonRpcCacheProvider` that handles the batching of calls and sends them across to our new api endpoint
- Add new endpoint that handles unbatching and caching calls and rebatching cache misses before making a batched call to infura. Reconstitutes all results and passes back to client
- Put caching functionality behind a feature toggle for testing on staging
- Falls back to previous custom batching provider `JsonRpcBatchProvider` when flag is disabled and when using hardhat in staging or on vercel preview deployment
- Standardises all infura eth_calls to use the MCD_Spot contract address in the from field. This is required for OSMS calls. But, it also means all read_only calls are consistent between users and the cache key currently includes the from field when generating a digest.
  
## How to test 🧪
Find `pages/api/infuraCallsCache.tsx` and update line 24: `const cache = new BatchCache()` to this ` const cache = new BatchCache({ debug: true })`. Cache TTL can also be increased here to aid test
Toggle the feature flag to true for `BatchCache in your browser console.

- Navigate to various marketing pages - check network tab and see infura calls - at the same time open 2 other tabs in other browsers and request the same page in those browsers. You should see cache hits appearing in the console assuming requests from different tabs are arriving within the 15s default window.
- Open a multiply vault & earn vault - confirm all positions were opened without issue

---- ---- ----
**Testing with multiple browsers**

**1 Concurrent Sessions**
<img width="635" alt="Screenshot 2022-07-22 at 13 09 19" src="https://user-images.githubusercontent.com/5372463/180436432-11886c29-a411-4c5d-b8b0-4d298d68bcbd.png">

**2 Concurrent Sessions**
<img width="917" alt="Screenshot 2022-07-22 at 13 09 32" src="https://user-images.githubusercontent.com/5372463/180436475-ee22d5e4-66d3-4d38-be70-92b7fa0e7c1f.png">

**2 + Additional 3 Sessions within 15 seconds TTL window**
<img width="1122" alt="Screenshot 2022-07-22 at 13 09 45" src="https://user-images.githubusercontent.com/5372463/180436567-de415bfc-36ed-4d04-a58b-1fd2b53abb4f.png">
